azurelinuxagent/common/protocol/util.py

```diff
@@ -216,82 +188,51 @@ def _clear_wireserver_endpoint(self):
-
-        try:
-            protocol = WireProtocol(endpoint)
-            protocol.detect()
-            self._set_wireserver_endpoint(endpoint)
-            return protocol
-        except ProtocolError as e:
-            logger.info("WireServer is not responding. Reset dhcp endpoint")
-            self.dhcp_handler.endpoint = None
-            self.dhcp_handler.skip_cache = True
-            raise e
-
-    def _detect_metadata_protocol(self):
-        protocol = MetadataProtocol()
-        protocol.detect()
-        return protocol
-
-    def _detect_protocol(self, protocols):
+    def _detect_protocol(self):
         """
         Probe protocol endpoints in turn.
         """
         self.clear_protocol()
 
         for retry in range(0, MAX_RETRY):
-            for protocol_name in protocols:
-                try:
-                    protocol = self._detect_wire_protocol() \
-                                if protocol_name == prots.WireProtocol \
-                                else self._detect_metadata_protocol()
+            try:
+                endpoint = self.dhcp_handler.endpoint
+                if endpoint is None:
+                    '''
+                    Check if DHCP can be used to get the wire protocol endpoint
+                    '''
+                    dhcp_available = self.osutil.is_dhcp_available()
+                    if dhcp_available:
+                        logger.info("WireServer endpoint is not found. Rerun dhcp handler")
+                        try:
+                            self.dhcp_handler.run()
+                        except DhcpError as e:
+                            raise ProtocolError(ustr(e))
+                        endpoint = self.dhcp_handler.endpoint
+                    else:
+                        logger.info("_detect_protocol: DHCP not available")
+                        endpoint = self.get_wireserver_endpoint()
 
-                    return (protocol_name, protocol)
+                try:
+                    protocol = WireProtocol(endpoint)
+                    protocol.detect()
+                    self._set_wireserver_endpoint(endpoint)
+                    return protocol
 
                 except ProtocolError as e:
-                    logger.info("Protocol endpoint not found: {0}, {1}",
-                                protocol_name, e)
+                    logger.info("WireServer is not responding. Reset dhcp endpoint")
+                    self.dhcp_handler.endpoint = None
+                    self.dhcp_handler.skip_cache = True
+                    raise e
+    
+            except ProtocolError as e:
+                logger.info("Protocol endpoint not found: {0}", e)
 
             if retry < MAX_RETRY - 1:
-                logger.info("Retry detect protocols: retry={0}", retry)
+                logger.info("Retry detect protocol: retry={0}", retry)
                 time.sleep(PROBE_INTERVAL)
         raise ProtocolNotFoundError("No protocol found.")
```

Comment:
So - It can wait - in theory- for 1 hour before failing to raise ProtocolNotFoundError()? Just confirming.

I understand Telemetry events won't work either as telemetry requires the protocol to be enabled, for the Wireserver to receive the `/telemetrydata` request

Reply:
Is there guidance on how much we should wait? or we decide the wait time period.

Reply:
I'll investigate and update if reqd.

Reply:
@byue - Dont worry about this comment.